### DISCUSSION TITLE
lgtm画像のフォントのカスタマイズを実装する調査

### DISCUSSION THREAD

**Discussion by kazzyfrog** [lgtm-factory/lgtm-factory#130] [Scrap]
lgtm画像のフォントのカスタマイズを実装する調査を、つらつら書いていきますー

下記が、すでに参考になりそう！
- https://trpfrog.net/blog/google-fonts-on-satori
- #70

**Comment by kazzyfrog** [comment]
Google Fonts の Developer API？からフォントデータを取得する実装なのかな。。

**Comment by kazzyfrog** [comment]
https://trpfrog.net/blog/google-fonts-on-satori
上記の記事を参考に、Google Fonts の Developer API？からフォントデータを取得する実装が可能っぽい！把握！！


そして、以下より、
- https://github.com/lgtm-factory/lgtm-factory/discussions/70#discussioncomment-10379858

> > プロジェクトにフォントデータを置かなくてもgoogle fonts使えるようになりました！
> 
> って書いたけど、ローカルで開発＆確認するときはenvファイルに自分でgoogle api key取得してきてセットしないといけないから、難しい？面倒？かな・・・
> それよりgoogle fontsのサイトからDLしてassetsファイルに配置した方が、楽かな？
> でも、それをたくさんやるとサイトが重くなりそう、ビルドとか大丈夫なのだろうか・・・


なるほど、、、確かにその通りだ。。
作成時に、ローカルでフォント確認できないのつらいかもですが、他に方法はなさそう、、かな。。
少し調べる

**Comment by kazzyfrog** [comment]
多分だけど、V2のAPIなら、アクセスキーなしで叩けそう！
https://developers.google.com/fonts/docs/css2?hl=ja

１つだけ指定したらできた！
satoriのplaygroundも、V2の方で、多分アクセスキーなしでfont API叩いている！

明日いろいろ、動的なフォントで読み込めるか検証する！！

**Reply by kazzyfrog** [reply]
V2で多分いけた！

1. Google Fontのサイトから、好きなフォントを選び、太さ（weight）を選択する
2. 「Get embed code」ボタンをクリックし、フォントのURLをコピー：`https://fonts.googleapis.com/css2?family=Caveat:wght@700`
<img width="443" alt="スクリーンショット 2024-08-28 1 10 10" src="https://github.com/user-attachments/assets/f9b82d80-9108-4c0c-b1d5-b251909c14cf">

3. 上記のURLのレスポンスから、実際のバイナリーデータのあるURLを取得`https://fonts.gstatic.com/s/caveat/v18/WnznHAc5bAfYB2QRah7pcpNvOx-pjRV6SII.ttf`
4. `const fontData = await response.arrayBuffer()`で、オプションのfontDataに渡す！

**Reply by kazzyfrog** [reply]
ポイント：
- グーグルフォントAPIのV２のURLは、`https://fonts.googleapis.com/css`で始まる
- `?family=Caveat:wght@700`クリパラメータを使用して、フォントとweightを指定する

詳細は、公式（グーグルフォントAPIのV２）：https://developers.google.com/fonts/docs/css2?hl=ja

**Reply by kazzyfrog** [reply]
なので、UXとしては、
- ユーザーにフォント決めてもらって、Google Fontのサイトから、URLをコピーしてもらう！
- 形式は、`https://fonts.googleapis.com/css2?family=Caveat:wght@700`この形！

**Reply by kazzyfrog** [reply]
CSSファイルから、ぶっこ抜くという荒技を、satori公式はしていたので、真似た！

**Comment by kagomen** [comment]
おおおーー！！！
すごい！！これでフォント選び放題！！
調査ありがとうございますー！！！😻

**Comment by kazzyfrog** [comment]
メモ：
日本語フォントがやっかい。。😈
Next.js でも、フォントデータはデカいのでキャッシュされないことがある(平気で2MB を超える)

グーグルフォントのクエリパラメータを使用して、`text=おはよう`などで、必要な分だけ読み込めばめっちゃ軽くなって、キャッシュも効く

**Reply by kazzyfrog** [reply]
まあ一旦パフォーマンスは無視！
気になった時に修正する！
